lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.watch.TimestampToVersionMap.CommitInfo;
+import com.palantir.lock.watch.TimestampToVersionMap.MapEntry;
+import com.palantir.logsafe.Preconditions;
+
+/**
+ * This class should only be used through {@link FailureCheckingLockWatchEventCache} as a proxy; failure to do so will
+ * result in concurrency issues and inconsistency in the cache state.
+ */
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final TimestampToVersionMap timestampMap;
```

Comment:
Entries are only ever removed when a certain method is called, which is called on `finish`. However, we do not _guarantee_ that a user calls `finish`; as such, there is a potential case where throwaway transactions are created, added to this map, but never deleted.

How do we want to go about handling this case? We could implement a retention policy but would that be a problem for long-running transactions (or do we care about those)?

Reply:
Or is this just a matter of letting people be responsible for calling `finish` on their transactions? I feel like that is an approach prone to memory leaks.

Reply:
So, 2 pieces:

1. We need to make sure that this cleanup always happens when users are using #runWithRetry. I don't think this is happening right now, so needs a fixup.
2. For anyone using OpenTransaction, we need to rely on them calling #finish or whatever else we decide is the way to do it.

Reply:
Fixed.